include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1466,6 +1466,119 @@ struct __partial_merge_kernel
     }
 };
 
+//Searching for an intersection of virtual matrix (n1, n2) diagonal with the Merge Path to define sub-ranges
+//to serial merge. For example, a virtual matrix for [0,1,1,2,3] and [0,0,2,3] shown below:
+//     0   1  1  2   3
+//    ------------------
+//   |--->
+// 0 | 0 | 1  1  1   1
+//   |   |
+// 0 | 0 | 1  1  1   1
+//   |   ---------->
+// 2 | 0   0  0  0 | 1
+//   |             ---->
+// 3 | 0   0  0  0   0 |
+template<typename Rng1, typename Rng2, typename _Index, typename _Index1, typename _Index2>
+auto __find_start_point(Rng1& __rng1, Rng2& __rng2, _Index __i_elem, _Index1 __n_1, _Index2 __n_2)
+{
+    _Index1 __end1 = 0;
+    _Index2 __end2 = 0;
+    if (__i_elem < __n_1) //a condition to specify upper or down part of the merge matrix to be processed
+    {
+        auto __q = __i_elem; //digonal index
+        auto __n_diag = __q < __n_2 ? __q : __n_2; //digonal size
+
+        oneapi::dpl::counting_iterator<_Index> __diag_it(0);
+        auto __res = std::lower_bound(__diag_it, __diag_it + __n_diag, 1, [&__rng1, &__rng2, __q](auto __v1, auto __v2)
+            {
+                auto __val1 = __rng1[__q - __v1 - 1] < __rng2[__v1];
```

Comment:
`__val1` is a bool as it is the result of the `<` comparison, then we compare it to the diagonal index `__v2`?

Is there a better name for `__val1` to indicate what it is supposed to represent?

Reply:
'__v2' is not diagonal index - this is a value to be found. We are searching for '1'.

Reply:
thanks, that helps.

Reply:
Actually, I'm still not convinced about `__v2`.  Is it not just the next element of the `counting_iterator`?

It seems like you are wanting the input iterator sequence to be already transformed into  `__rng1[__q - idx - 1] < __rng2[idx]` where `idx` is  is the `counting_iterator` element, but that isn't the case as written.  

I think you would need to zip together the input ranges, have a transform view and a functor which applies the above expression, then pass in that transform iterator into the `lower_bound` algorithm to get the result you are looking for here.

Unless I'm missing something, `__v2` is just the next element of the input counting iterator after `__v1` and it has not been similarly transformed.

Reply:
Cppreference tells about std::lower_bound:
```
Returns an iterator pointing to the first element in the range [first, last) such that element < value (or comp(element, value)) is false, (i.e. that is greater than or equal to value), or last if no such element is found.

The range [first, last) must be partitioned with respect to the expression element < value (or comp(element, value)), i.e., all elements for which the expression is true must precede all elements for which the expression is false. A fully-sorted range meets this criterion.
```

Reply:
Sorry, I've been thinking about it wrong.  My mistake.